src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -336,16 +546,110 @@ private BoundExpression RewriteStringConcatenationFourExprs(SyntaxNode syntax, B
+
+            var stringifiedArgs = ArrayBuilder<BoundExpression>.GetInstance(loweredArgs.Length);
+
+            foreach (var loweredArg in loweredArgs)
+            {
+                Debug.Assert(loweredArg.HasErrors || loweredArg.Type is { } argType && (argType.IsStringType() || argType.IsCharType()));
+
+                if (loweredArg.Type?.IsCharType() == true)
+                {
+                    stringifiedArgs.Add(WrapCharExprIntoToStringCall(loweredArg));
+                }
+                else
+                {
+                    stringifiedArgs.Add(loweredArg);
+                }
+            }
 
             var method = UnsafeGetSpecialTypeMethod(syntax, SpecialMember.System_String__ConcatStringArray);
             Debug.Assert((object)method != null);
 
-            var array = _factory.ArrayOrEmpty(_factory.SpecialType(SpecialType.System_String), loweredArgs);
+            var array = _factory.ArrayOrEmpty(_factory.SpecialType(SpecialType.System_String), stringifiedArgs.ToImmutableAndFree());
 
             return BoundCall.Synthesized(syntax, receiverOpt: null, initialBindingReceiverIsSubjectToCloning: ThreeState.Unknown, method, array);
         }
 
+        private BoundExpression WrapCharExprIntoToStringCall(BoundExpression expr)
+        {
+            var exprType = expr.Type;
+            Debug.Assert(exprType is not null && exprType.IsCharType());
+
+            var objectToString = UnsafeGetSpecialTypeMethod(expr.Syntax, SpecialMember.System_Object__ToString);
+            var charToString = FindSpecificToStringOfStructType(exprType, objectToString);
+
+            if (charToString is not null && !IsFieldOfMarshalByRef(expr))
+            {
+                return BoundCall.Synthesized(expr.Syntax, expr, initialBindingReceiverIsSubjectToCloning: ThreeState.Unknown, charToString);
+            }
+
+            return expr;
+        }
+
+        private BoundExpression WrapIntoReadOnlySpanOfCharIfNecessary(BoundExpression expr, MethodSymbol stringImplicitConversionToReadOnlySpan, MethodSymbol readOnlySpanCtorRefParamChar, out LocalSymbol? tempLocal)
+        {
+            tempLocal = null;
+            var exprType = expr.Type;
+
+            Debug.Assert(exprType is not null && (exprType.IsStringType() || exprType.IsCharType() || exprType.IsReadOnlySpanChar()));
+
+            if (expr.Type.IsReadOnlySpanChar())
+            {
+                return expr;
+            }
+            else if (exprType.IsStringType())
+            {
+                return BoundCall.Synthesized(expr.Syntax, receiverOpt: null, initialBindingReceiverIsSubjectToCloning: ThreeState.Unknown, stringImplicitConversionToReadOnlySpan, expr);
+            }
+            else
+            {
+                if (expr is BoundParameter { ParameterSymbol.RefKind: RefKind.None } or BoundLocal { LocalSymbol.RefKind: RefKind.None })
+                {
+                    // new ReadOnlySpan<char>(in <expr>)
```

Comment:
>// new ReadOnlySpan<char>(in <expr>) [](http://example.com/codeflow?start=20&length=36)

Is this really safe? What if some following operand for the concatenation is changing value of the local/parameter?

Reply:
Unfortunatelly, it is not safe. Added tests to verify this case and removed code path with taking direct reference of a local. It is sad to see how codegen degrades in most common scenarios to satisfy such tiny edge cases(